swarm/pss/prox_test.go

```diff
@@ -255,93 +243,153 @@ func testProxNetwork(t *testing.T) {
+		case hn := <-d.msgC:
+			received++
+			log.Debug("msg received", "msgs_received", received, "total_expected", d.requiredMessages, "id", hn.id, "serial", hn.serial)
+		}
+	}
+
+	d.setDone()
+	time.Sleep(time.Millisecond * 16) // allow the nodeMsgHandlers to complete if running
+	close(d.doneC)
+
 	for {
 		select {
-		case err := <-tstdata.errC:
-			return err
-		case hn := <-tstdata.msgC:
+		case <-d.errC:
+			// channel is closed, now function can return
+			return res
+		case <-d.msgC:
 			received++
-			log.Debug("msg received", "msgs_received", received, "total_expected", tstdata.requiredMessages, "id", hn.id, "serial", hn.serial)
-			if received == tstdata.allowedMessages {
-				close(tstdata.doneC)
-				return nil
-			}
 		}
 	}
-	return nil
+	return res
 }
 
-func handlerChannelListener(tstdata *testData, ctx context.Context) {
+func handlerChannelListener(d *testData) {
 	for {
 		select {
-		case <-tstdata.doneC: // graceful exit
-			tstdata.setDone()
-			tstdata.errC <- nil
-			return
-
-		case <-ctx.Done(): // timeout or cancel
-			tstdata.setDone()
-			tstdata.errC <- ctx.Err()
+		case <-d.doneC: // graceful exit
+			close(d.errC)
 			return
 
 		// incoming message from pss message handler
-		case handlerNotification := <-tstdata.handlerC:
+		case handlerNotification := <-d.handlerC:
 			// check if recipient has already received all its messages and notify to fail the test if so
-			aMsgs := tstdata.allowedMsgs[handlerNotification.id]
-			if len(aMsgs) == 0 {
-				tstdata.setDone()
-				tstdata.errC <- fmt.Errorf("too many messages received by recipient %x", handlerNotification.id)
-				return
+			h := handlerNotification.id
+			if len(d.allowedMsgs[h]) == 0 {
+				d.errC <- fmt.Errorf("too many messages received by recipient %x", handlerNotification.id)
+				break
 			}
 
 			// check if message serial is in expected messages for this recipient and notify to fail the test if not
 			idx := -1
-			for i, msg := range aMsgs {
-				if handlerNotification.serial == msg {
+			for i, s := range d.allowedMsgs[h] {
+				if handlerNotification.serial == s {
 					idx = i
 					break
 				}
 			}
 			if idx == -1 {
-				tstdata.setDone()
-				tstdata.errC <- fmt.Errorf("message %d received by wrong recipient %v", handlerNotification.serial, handlerNotification.id)
-				return
+				d.errC <- fmt.Errorf("message %d received by wrong recipient %v", handlerNotification.serial, handlerNotification.id)
+				break
 			}
 
-			// message is ok, so remove that message serial from the recipient expectation array and notify the main sim thread
-			aMsgs[idx] = aMsgs[len(aMsgs)-1]
-			aMsgs = aMsgs[:len(aMsgs)-1]
-			tstdata.msgC <- handlerNotification
+			// message is ok, so remove that message serial from the recipient expectation array
+			last := len(d.allowedMsgs[h]) - 1
+			d.allowedMsgs[h][idx] = d.allowedMsgs[h][last]
+			d.allowedMsgs[h] = d.allowedMsgs[h][:last]
+			//notify the main sim thread
+			d.msgC <- handlerNotification
```

Comment:
This channel is unbuffered.  If there is no reader it will block.
So if the test timeouts we might leak this goroutine as I'm not sure we dry out `d.msgC` before exiting the test.

Do we know what amount of messages will go down on `d.msgC` prior to creating the channel?